Classes/Event/EventPublisher.php

```diff
@@ -45,46 +47,65 @@ class EventPublisher
     {
         $this->eventStore = $eventStore;
-        $this->eventBus = $eventBus;
+        $this->eventListenerLocator = $eventListenerLocator;
         $this->propertyMapper = $propertyMapper;
         $this->eventTypeResolver = $eventTypeResolver;
     }
 
 
     /**
      * @param string $streamName
-     * @param EventWithMetadata $eventWithMetadata
+     * @param EventInterface $event
      * @param int $expectedVersion
      */
-    public function publish(string $streamName, EventWithMetadata $eventWithMetadata, int $expectedVersion = ExpectedVersion::ANY)
+    public function publish(string $streamName, EventInterface $event, int $expectedVersion = ExpectedVersion::ANY)
     {
-        $this->publishMany($streamName, [$eventWithMetadata], $expectedVersion);
+        $this->publishMany($streamName, [$event], $expectedVersion);
     }
 
     /**
      * @param string $streamName
-     * @param EventWithMetadata[] $eventsWithMetadata
+     * @param EventInterface[] $events
      * @param int $expectedVersion
      */
-    public function publishMany(string $streamName, array $eventsWithMetadata, int $expectedVersion = ExpectedVersion::ANY)
+    public function publishMany(string $streamName, array $events, int $expectedVersion = ExpectedVersion::ANY)
     {
         $convertedEvents = new WritableEvents();
-        foreach ($eventsWithMetadata as $eventWithMetadata) {
-            $type = $this->eventTypeResolver->getEventType($eventWithMetadata->getEvent());
-            $data = $this->propertyMapper->convert($eventWithMetadata->getEvent(), 'array');
-            $metadata = $this->propertyMapper->convert($eventWithMetadata->getMetadata(), 'array');
+        foreach ($events as $event) {
+            $type = $this->eventTypeResolver->getEventType($event);
+            $metadata = [];
+            $this->emitPublishEvent($event, $metadata);
+            $data = $this->propertyMapper->convert($event, 'array');
             $convertedEvents->append(new WritableEvent($type, $data, $metadata));
         }
-        $this->eventStore->commit($streamName, $convertedEvents, $expectedVersion);
-        foreach ($eventsWithMetadata as $eventWithMetadata) {
-            $this->eventBus->handle($eventWithMetadata);
+        $storedEvents = $this->eventStore->commit($streamName, $convertedEvents, $expectedVersion);
+
+        $configuration = new PropertyMappingConfiguration();
+        $configuration->allowAllProperties();
+        $configuration->forProperty('*')->allowAllProperties();
+        foreach ($storedEvents as $storedEvent) {
+            $eventClassName = $this->eventTypeResolver->getEventClassNameByType($storedEvent->getType());
+            $event = $this->propertyMapper->convert($storedEvent->getPayload(), $eventClassName, $configuration);
+            foreach ($this->eventListenerLocator->getListeners($storedEvent->getType()) as $listener) {
+                call_user_func($listener, $event, $storedEvent);
+            }
```

Comment:
This is one detail I struggle with:
The problem is not that we convert the event twice here (I think that's actually a good thing as we then use the exact same mechanism during initial handling and replay and thus can fail early if sth is wrong)
But that code is now duplicated in `EventPublisher::publishMany()` and `EventStream::current()`